sdk/cosmos/azure-cosmos/azure/cosmos/_global_partition_endpoint_manager_per_partition_automatic_failover.py

```diff
@@ -0,0 +1,263 @@
+        self._threshold_lock = threading.Lock()
+
+    def is_per_partition_automatic_failover_enabled(self) -> bool:
+        if not self._database_account_cache or not self._database_account_cache._EnablePerPartitionFailoverBehavior:
+            return False
+        return True
+
+    def is_per_partition_automatic_failover_applicable(self, request: RequestObject) -> bool:
+        if not self.is_per_partition_automatic_failover_enabled():
+            return False
+
+        if not request:
+            return False
+
+        if (self.location_cache.can_use_multiple_write_locations_for_request(request)
+                or _OperationType.IsReadOnlyOperation(request.operation_type)):
+            return False
+
+        # if we have at most one region available in the account, we cannot do per partition automatic failover
+        available_regions = self.compute_available_preferred_regions(request)
+        if len(available_regions) <= 1:
+            return False
+
+        # if the request is not a non-query plan document request
+        # or if the request is not executing a stored procedure, return False
+        if (request.resource_type != ResourceType.Document and
+                request.operation_type != _OperationType.ExecuteJavaScript):
+            return False
+
+        return True
+
+    def try_ppaf_failover_threshold(
+            self,
+            pk_range_wrapper: "PartitionKeyRangeWrapper",
+            request: "RequestObject"):
+        """Verifies whether the per-partition failover threshold has been reached for consecutive errors. If so,
+        it marks the current region as unavailable for the given partition key range, and moves to the next available
+        region for the request.
+
+        :param PartitionKeyRangeWrapper pk_range_wrapper: The wrapper containing the partition key range information
+            for the request.
+        :param RequestObject request: The request object containing the routing context.
+        :returns: None
+        """
+        # If PPAF is enabled, we track consecutive failures for certain exceptions, and only fail over at a partition
+        # level after the threshold is reached
+        if request and self.is_per_partition_automatic_failover_applicable(request):
+            if (self.ppaf_thresholds_tracker.get_pk_failures(pk_range_wrapper)
+                    >= int(os.environ.get(Constants.TIMEOUT_ERROR_THRESHOLD_PPAF,
+                                          Constants.TIMEOUT_ERROR_THRESHOLD_PPAF_DEFAULT))):
+                # If the PPAF threshold is reached, we reset the count and mark the endpoint unavailable
+                # Once we mark the endpoint unavailable, the PPAF endpoint manager will try to move to the next
+                # available region for the partition key range
+                with self._threshold_lock:
+                    logger.warning("PPAF - Failover threshold reached for partition key range: %s", pk_range_wrapper)
```

Comment:
Failover from / to which region along with `pk_range_wrapper` instance can add better context to L137.